### PR TITLE
Handle arguments in Python frontend

### LINKED ISSUE #1653: Python frontend: better argument handling
The Python frontend is missing some functionality when handling arguments:
- [x] `vararg` https://github.com/Fraunhofer-AISEC/cpg/blob/06995d44afb21212c72fe18320324d76d562575a/cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt#L438
- [ ] `kwonlyargs` https://github.com/Fraunhofer-AISEC/cpg/blob/06995d44afb21212c72fe18320324d76d562575a/cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt#L448
- [ ] `kw_defaults` https://github.com/Fraunhofer-AISEC/cpg/blob/06995d44afb21212c72fe18320324d76d562575a/cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt#L458
- [ ] `kwarg` https://github.com/Fraunhofer-AISEC/cpg/blob/06995d44afb21212c72fe18320324d76d562575a/cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt#L468
- [ ] `defaults` https://github.com/Fraunhofer-AISEC/cpg/blob/06995d44afb21212c72fe18320324d76d562575a/cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt#L478

### PR DESCRIPTION
**Pull request by lshala**
This partially fixes #1653

### CODE REVIEW

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:428]: Change to `result.additionalProblems += ...` once https://github.com/Fraunhofer-AISEC/cpg/pull/1663/ is merged.

**Comment by oxisto** [comment]
Can you please merge/rebase main, thanks!

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:432]: I don't think this is correct: assume there are some default values, then this will always get a default value and you'll create an error a few lines below -> this needs some fixing.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:453]: Please also update to a Problem node, once the other PR is in main.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:408]: is a keyword arg really variadic?

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:413]: doc string

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:458]: doc string

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:484]: doc string

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:567]: doc string

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:466]: This indexing math looks off to me. Assume `foo(self, a, b = 2, c = 3):` then
- `nonDefaultArgsCount = 4 - 2 = 2`
- `positionalArguments.drop(startIdx)` => this will drop the receiver => a list of 3 elements with indices 0,1,2
- `defaultIndex = idx - nonDefaultArgsCount = 0 - 2 = -2` ???

Either I'm confused or this is off.

**Review by maximiliankaul — APPROVED** [review]
I like this PR :+1: as it significantly improves the code quality / readability of my messy argument handling code :see_no_evil: 
There are some nitpicking points to be taken care of and the indexing needs another look (maybe I'm wrong - you're tests seem to work)... I'm confused -> will look at this after my vacation if no one else approves by then.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:478]: is this `isKwoOnly = defaultValue != null,` obvious? otherwise: please some doc

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:466]: I just tried this with the debugger -> there is an error as this code does not find the default for `b`.
Please also include a test case for:
```Python
class Foo:
  def foo(self, a, b = 1, c = 2):
    pass
```

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt]: Done

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt]: Done

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt]: Done

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:379]: Done

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:423]: Done

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt]: Done

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:553]: Done

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:443]: Since the property “isKwoOnly” is a boolean expression, the result i.e. if the condition is true, returns true otherwise false

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:392]: At the moment we are not handling the case that a receiver can have a default value. As discussed, the receiver (if available) will be created as a variable. I added this information to the doc string.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:431]: The indexing math ensures that default arguments are applied correctly by first dropping the receiver. The defaultIndex is calculated by subtracting the number of non-default arguments, which might result in negative values for arguments without a default. However, this is intentional because we only use defaultIndex when it's greater than or equal to 0, meaning it only applies to arguments with defaults.

I added another test with self, which like the others also passes.

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt
Comment (KuechA) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt:51]: Isn't there `kotlin.test.BeforeClass`? This could make it a bit nicer and get rid of the companion and JvmStatic annotation

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt
Comment (KuechA) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt:137]: Does it make sense to add a check that there are no additional arguments (e.g. by checking the size of `func.parameters` is as expected)?

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt
Comment (KuechA) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt:112]: Does it make sense to add a check that there are no additional arguments (e.g. by checking the size of `func.parameters` is as expected)?

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:431]: Please see this test https://github.com/Fraunhofer-AISEC/cpg/pull/1669/commits/8b15388ed69913a33abbe6e48d37efe5b09cecc2 (feel free to edit / update it as I currently do not have access to a Java IDE).

Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:431]: Also: it would be great to have your answer in the code as to get a better understanding what and why things are happening here. Maybe a URL to the Python docs where the argument handling is introduced?

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:443]: I'm still confused. The method is called `handlePositionalArguments` and this code reads like the _positional_ argument can be _keyword only_? I need to read up on this.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:553]: I fail to see it. Is this some Java / Intellij magic again?

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt
Comment (maximiliankaul) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt:68]: Some nitpicking: can we make constant stuff `val` instead of `var`?

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (KuechA) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:443]: Thinking more about it: Don't the keyword and positional arguments affect the function calls more than the declarations? Does this "only" require fixing the SymbolResolver or also an update of the frontend?

I think that I can always use the keyword syntax even for positional arguments. Maybe the "only" makes the difference here which is somewhat confusing and it should be "isKeyword" and "isPositional" and then, there might be a combination where everything can be used in the function call

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt
Comment (lshala) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt]: good point. Done

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt
Comment (lshala) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt]: Done

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt
Comment (lshala) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt]: Done

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:591]: Sorry, my bad. 
Done.

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt
Comment (lshala) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler/ArgumentsHandlerTest.kt]: I've changed it to the suggested one.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt]: As we are handling not only the posonlyargs but also the "normal" args in this method and these can either be kwonlyargs or not, I removed the property isKwOnly here.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:462]: I refactored the method and also added a doc string. Thank you for adding an additional test. Now this test also passes.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt
Comment (KuechA) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:414]: Ok, but I think if you have something like `def foo(self, a, b=1)` you would currently get the default value `1` here. I think you have to fix the "offset" of the first default value like in the method below (`positionalArguments.size - args.defaults.size`) or so.

Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt]: Done

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/statementHandler/StatementHandlerTest.kt
Comment (oxisto) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/statementHandler/StatementHandlerTest.kt:54]: Personally I don't like this pattern, but I won't block this PR any further. What is the advantage of storing the result in the class? I only see this pattern usually if multiple tests execute on the same file, but each test is executed on another file and the `result` property gets overwritten. This feels very error prone. I get that the tests contain a lot of boilerplate but then we should see how we can rewrite the `analyze` functions to be leaner and still return the result/tu instead of assigning it to a class member.

Comment (lshala) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/statementHandler/StatementHandlerTest.kt:54]: I totally agree with you. I used this pattern for the tests in ArgumentsHandlerTest.kt, in which we are using the same file for every test. 
I will improve it over time.

**Pull request merged**
